src/data/simple_csr_source.cu

```diff
@@ -12,102 +12,212 @@
+    if ((column.valid.Data() == nullptr || column.valid.Check(tid)) &&
+        !common::CheckNAN(column.data[tid])) {
+      AssignValue(column.data[tid], colid, offsets, out_data);
+    }
+  } else if (missing_is_nan) {
+    // specified missing value, but it's NaN
+    if (!common::CheckNAN(column.data[tid])) {
+      AssignValue(column.data[tid], colid, offsets, out_data);
+    }
+  } else {
+    // specified missing value, and it's not NaN
+    if (!common::CloseTo(column.data[tid], missing)) {
+      AssignValue(column.data[tid], colid, offsets, out_data);
+    }
   }
 }
 
-void SimpleCSRSource::FromDeviceColumnar(std::vector<Columnar> cols) {
-  uint64_t const n_cols = cols.size();
-  uint64_t const n_rows = cols[0].size;
+template <typename T>
+void FromDeviceColumnarImpl(std::vector<Json> const& columns,
+                            bool has_missing, float missing,
+                            SparsePage* p_page, MetaInfo* p_info) {
+  size_t const n_columns = columns.size();
+  std::vector<Columnar<T>> foreign_cols(n_columns);
+
+  for (size_t i = 0; i < foreign_cols.size(); ++i) {
+    CHECK(IsA<Object>(columns[i]));
+    auto const& column = get<Object const>(columns[i]);
+    foreign_cols[i] = ArrayInterfaceHandler::ExtractArray<T>(column);
+  }
+
+  p_info->num_col_ = n_columns;
+  p_info->num_row_ = foreign_cols[0].size;
 
-  auto ptr = cols[0].data.data();
+  uint64_t const n_cols = foreign_cols.size();
+  uint64_t const n_rows = foreign_cols[0].size;
+
+  auto ptr = foreign_cols[0].data.data();
   int32_t device = dh::CudaGetPointerDevice(ptr);
   CHECK_NE(device, -1);
 
   for (int32_t i = 1; i < n_cols; ++i) {
-    auto ptr = cols[i].data.data();
+    auto ptr = foreign_cols[i].data.data();
     int32_t ptr_device = dh::CudaGetPointerDevice(ptr);
     CHECK_EQ(device, ptr_device)
         << "GPU ID at 0^th column: " << device << ", "
         << "GPU ID at column " << i << ": " << ptr_device;
+    CHECK_EQ(foreign_cols[0].size, foreign_cols[i].size)
+        << "Columns should be of same size.";
   }
 
   dh::safe_cuda(cudaSetDevice(device));
 
-  page_.offset.SetDevice(device);
-  page_.offset.Resize(info.num_row_ + 1);
-
-  page_.data.SetDevice(device);
-  page_.data.Resize(info.num_nonzero_);
+  p_page->offset.SetDevice(device);
+  p_page->offset.Resize(p_info->num_row_ + 1);
 
-  auto s_data = page_.data.DeviceSpan();
-  auto s_offsets = page_.offset.DeviceSpan();
+  auto s_offsets = p_page->offset.DeviceSpan();
   CHECK_EQ(s_offsets.size(), n_rows + 1);
 
   int32_t constexpr kThreads = 256;
-  dh::device_vector<Columnar> d_cols(cols);
+  dh::device_vector<Columnar<T>> d_cols(foreign_cols);
   auto s_d_cols = dh::ToSpan(d_cols);
 
   dh::safe_cuda(cudaMemset(s_offsets.data(), 0, sizeof(int32_t) * (n_rows + 1)));
-
-  CountValidKernel<kThreads><<<n_cols, kThreads>>>(s_d_cols, n_rows, s_offsets);
+  dh::caching_device_vector<int32_t> d_flag;
+  if (!common::CheckNAN(missing)) {
+    d_flag.resize(1, 0);
```

Comment:
dh::caching_device_vector has a no-op constructor so I'm not sure if these values are actually being initialised on resize  (see https://github.com/dmlc/xgboost/blob/c89bcc4de5368b3f8a7fa170d8348287dab44caf/src/common/device_helpers.cuh#L339).

This should probably be documented much better. Another way to do this would be initialise the vector with size 1, then call thrust::fill.